packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -781,6 +795,17 @@ class RenderParagraph extends RenderBox
     return _textPainter.getWordBoundary(position);
   }
 
+  /// Returns the text range of the line at the given offset.
+  ///
+  /// The newline, if any, is included in the range.
+  ///
+  /// Valid only after [layout].
+  TextRange getLineBoundary(TextPosition position) {
```

Comment:
The primary reason this is not exposed is API bloat, since once we add it, it cannot be removed. We should try our best to explore reasonable alternatives before resorting to adding additional API.

In this case, getLineBoundary exposes a few of the text position APIs based on metrics, which we want to eventually move away from to contain more bitdepth.

I think you may be able to workaround this by calling getBoxesForSelection, once with a collapsed selection to get the rect that bounds the current position, and again with a wide selection that encompasses a few lines. Then, you can take the right most and left most boundaries of the boxes that share the same y offset as the first call. Then pass those boundaries to getPositionForOffset.

That should get you something that may work for your purposes. It is a bit of a hassle, but I am hesitant to prematurely expose this particular API.

Reply:
Let me know if you are able to get a workaround working!

Reply:
Yep it does seem to work pretty well. Thanks for the suggestion! Since each RenderParagraph in Zefyr represents a single paragraph of text (meaning text between two newlines) here is what I ended up with:

https://github.com/memspace/zefyr/blob/78a5ac11542d39a45cc11cb904416dec0c443da7/packages/zefyr/lib/src/rendering/editable_text_line.dart#L294-L309

```dart
  @override
  TextRange getLineBoundary(TextPosition position) {
    // getOffsetForCaret returns top-left corner of the caret. To find all
    // selection boxes on the same line we shift caret offset by 0.5 of
    // preferredLineHeight so that it's in the middle of the line and filter out
    // boxes which do not include this offset on the Y axis.
    final caret = getOffsetForCaret(position);
    final lineDy = caret.translate(0.0, 0.5 * preferredLineHeight(position)).dy;
    final boxes = getBoxesForSelection(
        TextSelection(baseOffset: 0, extentOffset: node.length - 1));
    final lineBoxes = boxes
        .where((element) => element.top < lineDy && element.bottom > lineDy)
        .toList(growable: false);
    final start = getPositionForOffset(Offset(lineBoxes.first.left, lineDy));
    final end = getPositionForOffset(Offset(lineBoxes.last.right, lineDy));
    return TextRange(start: start.offset, end: end.offset);
  }
```

I'm checking at the vertical middle of the line because I've noticed that emojis, for instance, have larger selection boxes (and they also mess with the line height they are at):
![image](https://user-images.githubusercontent.com/441756/92557187-7db34f80-f220-11ea-89ee-5cc55227e273.png)

Overall it's actually pretty concise and I'm happy with this approach. So I removed `getLineBoundary` from this PR.

It now only exposes `getFullHeightForCaret`.